torch/csrc/jit/tensorexpr/loopnest.cpp

```diff
@@ -59,7 +59,7 @@ LoopNest::LoopNest(const std::vector<Tensor>& output_tensors) {
   verify(root_stmt_);
 }
 
-const std::unordered_set<BufPtr> LoopNest::getIntermediateBufs() const {
+std::unordered_set<BufPtr> LoopNest::getIntermediateBufs() const {
```

Comment:
I was at first surprised why we need this change but then found the answer in your second PR: we want to return mutable set so that `preallocate` function can erase preallocated bufs from it (is my understanding correct on this?). I wonder if a cleaner alternative would be to exclude buffers from `constants_` in this function (on line 67) - and to make this function stateless we can pass `constants_` to it as an optional parameter too. What do you think?

Reply:
> we want to return mutable set so that preallocate function can erase preallocated bufs from it (is my understanding correct on this?)

It's correct. This function collects all intermediate buffers into a set but the set should not be const so the user can add/delete buffers in the set. TEK modified the set (erased the buffers it pre-allocated) and passed it to `prepareForCodegen`. 

>  I wonder if a cleaner alternative would be to exclude buffers from constants_ in this function (on line 67) - and to make this function stateless we can pass constants_ to it as an optional parameter too
I am not sure if I understand you correctly. I don't think we can add filter (`constants_`) to this function from LoopNest. There are 2 call sites to this function: 1) TEK calls it to obtain the complete set of intermediate buffers and it does not need to exclude any buffers; 2) LoopNest::insertAllocFree calls it to obtain the set of intermediate buffers for runtime allocation and we need to exclude the buffers that have been preallocated in TEK (`constants_`) but how can we pass this info (`constants_`) to LoopNest from TEK?

Reply:
Right, passing `constants_` to LoopNest would be annoying.

How about the following crazy idea: what if we move logic of `prepareForCodegen` to the CodeGen constructor? In CodeGen we have access to BufferArgs, which essentially include inputs+outputs+constants, so it's trivial to determine which buffers need Alloc/Free. As a side benefit, we'll also move that logic to where it belongs better: to the codegen, and then we'll have less stateful methods in LoopNest, so hopefully we could completely get rid of it. What do you think?

It might be a sizable change, so if you'd like to land this version first I'd be ok with it, but then we'll have to at least add comments explaining why and what we pass, because currently the APIs become non-obvious.

Reply:
Yeah, currently intermediate buffer allocation (`insertAllocFree`) happens implicitly in LoopNest::prepareForCodegen and if we want to allow users to specify pre-allocation optimization in TEK we'll have to either 1) make `insertAllocFree` explicit so it's available to TEK, or 2) as you suggested, move `insertAllocFree` to codegen which happens after TEK.

I kinda like approach 2) at this point. Approach 1) requires some refactoring in LoopNest and it's not clear to me how to do it yet; approach 2) seems more neat since we only need to move `insertAllocFree` from LoopNest::prepareForCodegen to codegen constructor. 

And yeah, I'd love to explore approach 2) in another PR.  It seems to be an isolated task. I'll add comments for the intermediate buffer argument in this PR :)

Reply:
The PR looks good to me then! cc @bertmaher @navahgar for their opinions on (2)